accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,671 @@
+```
+
+### Casting
+Using arrays to help with the comparison (since function pointers are declared inline):
+```cs
+object arrInt = new delegate*<int>[1];
+var yikes = (delegate*<string>[])arrInt; // Expect InvalidCastException
+```
+which uses `castclass`
+```
+.locals init (object V_0, method string *()[] V_1)
+ldc.i4.1
+newarr     method int32 *()
+stloc.0
+ldloc.0
+castclass  method string *()[]
+stloc.1
+```
+### `Is` keyword
+```cs
+object arrInt = new delegate*<int>[1];
+bool f = arrInt is delegate*<string>[]; // 'false'
+```
+which uses `isinst`:
+```
+.locals init (object V_0, bool V_1)
+ldc.i4.1
+newarr     method int32 *()
+stloc.0
+ldloc.0
+isinst     method string *()[]
+ldnull
+cgt.un
+stloc.1
+```
+
+### `Is` keyword with known values
+C# will not use `isinst` when it "knows" the types are equal or not equal:
+```cs
+var fn = new delegate*<int>[1];
+Console.WriteLine(fn is delegate*<int>[]); // true
+Console.WriteLine(fn is delegate*<bool>[]); // false
+```
+```
+ldc.i4.1
+newarr     method int32 *()
+ldnull
+cgt.un
+call       void [System.Console]System.Console::WriteLine(bool)
+ldc.i4.0
+call       void [System.Console]System.Console::WriteLine(bool)
+```
+
+The C# compiler will also consider the `false` case warning CS0184: `The given expression is never of the provided...`.
+
+This C# vs. runtime disagreement with "known values" also appears with array variance:
+```cs
+Console.WriteLine(new int[1] is uint[]); // false
+Console.WriteLine(((object)new int[1]) is uint[]); // true
+```
+
+### With modifiers + unmodified types
+The above examples only compare parameter types without calling conventions or modifiers.
+
+An unmodified type does include managed vs. unmanaged as part of its type identity:
+```cs
+object o = new delegate*<int>[1];
+Console.WriteLine(o is delegate*<int>[]); // true
+Console.WriteLine(o is delegate* unmanaged[Cdecl]<int>[]); // false
+
+var fn = new delegate*<int>[1];
+Console.WriteLine(fn is delegate*<int>[]); // true
+Console.WriteLine(fn is delegate* unmanaged[Cdecl] <int>[]); // false
+```
+
+but unmanaged calling conventions are not part of the type identity:
+```cs
+object o1 = new delegate* unmanaged[SuppressGCTransition]<int>[1];
+Console.WriteLine(o1 is delegate* unmanaged[MemberFunction]<int>[]); // true
+
+object o2 = new delegate* unmanaged[CDecl]<int>[1];
+Console.WriteLine(o2 is delegate* unmanaged[StdCall]<int>[]); // true
+```
+
+In addition, all non-CallConv* modifiers, such as `ref\in\out\const` are not considered part of the type identity either.
+
+### With modified types
+A modified type is a simple wrapper that does not forward the other various `Type` members to the underlying type. It does, however intercept `GetFunctionPointerParameters()` to return modified type parameters. It also implements `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()` which throw on unmodified types.
```

Comment:
I'd consider `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()` to return `null` if it's not the special modified type. It matches how e.g. `GetElementType()` returns null on non-arrays.

Reply:
Returning `null` would make it different from the existing `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()` methods on `PropertyInfo`, `FieldInfo` and `MethodInfo`. I think it would be better to throw instead of returning null. People can check `IsModifiedType` property upfront if they are not sure whether they are dealing with modified type.

Reply:
The existing GetOptionalCustomModifiers on PropertyInfo, FieldInfo and MethodInfo return an empty array if there's no modifiers. Should we match that one?

Reply:
> The existing GetOptionalCustomModifiers on PropertyInfo, FieldInfo and MethodInfo return an empty array if there's no modifiers. Should we match that one?

I would prefer an empty array rather than `null`.

Reply:
> The existing GetOptionalCustomModifiers on PropertyInfo, FieldInfo and MethodInfo return an empty array if there's no modifiers. Should we match that one?

Yeah I think so. Throwing however may help users catch bugs (exception would say "only works on modified types"). Returning empty is preferred over `null` for consistency. So at this point I have a slight preference for returning empty for consistency vs. throwing.

Reply:
Any opinion on not whether we should try to forward all Type members (except for `Get*CustomModifiers` and `IsModifiedType`) to the underlying type to try to make it act like the underlying type? If we do that, then we probably should cache modified Type instances so the same instance is returned when everything matches including the custom mods. Caching however adds global memory for things that may only be used once and also may hamper assembly unloadability...

An option lightly discussed prior is to expose a non-Type instance like "SignatureMethod" mentioned in the design, but abandoned for various reasons. However, another approach is returning a wrapper for the custom mods like a "TypeModifiers" class:
```cs
public sealed class TypeModifiers
{
    public Type UnderlyingType { get; } // For access to fnptr parameter types etc.
    public Type[] GetOptionalCustomModifiers();
    public Type[] GetRequiredCustomModifiers();
}

// and returned by FieldInfo etc instead of System.Type
public abstract class FieldInfo
{
    public virtual Type? GetFieldTypeModifiers();
}
```
which makes the API more understandable IMO. i.e. no potential misuse of `Type.Get*CustomModifiers()`.